File: WordPressTVCore/Sources/WordPressTVCore/Data/HTML.swift
import Foundation

/// Minimal HTML-to-plain-text helper.
///
/// We can't use `NSAttributedString`'s HTML importer — it's unavailable on
/// tvOS — so this strips tags and decodes the handful of entities WP.com
/// titles and excerpts actually contain (named + numeric, decimal + hex).
enum HTML {
    static func plainText(_ raw: String) -> String {
        decodeEntities(stripTags(raw)).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func stripTags(_ s: String) -> String {
        s.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
    }

    static func decodeEntities(_ s: String) -> String {
        guard s.contains("&") else { return s }
        var result = ""
        var index = s.startIndex
        while index < s.endIndex {
            let char = s[index]
            if char == "&",
               let semicolon = s[index...].firstIndex(of: ";"),
               s.distance(from: index, to: semicolon) <= 12 {
                let body = String(s[s.index(after: index)..<semicolon])
                if let decoded = decode(entity: body) {
                    result.append(decoded)
                    index = s.index(after: semicolon)
                    continue
                }
            }
            result.append(char)
            index = s.index(after: index)
        }
        return result
    }

    /// `body` is the entity without the leading `&` or trailing `;`.
    private static func decode(entity body: String) -> Character? {
        if body.hasPrefix("#x") || body.hasPrefix("#X") {
            guard let code = UInt32(body.dropFirst(2), radix: 16),
                  let scalar = Unicode.Scalar(code) else { return nil }
            return Character(scalar)
        }
        if body.hasPrefix("#") {
            guard let code = UInt32(body.dropFirst()),
                  let scalar = Unicode.Scalar(code) else { return nil }
            return Character(scalar)
        }
        return named[body]
    }

    private static let named: [String: Character] = [
        "amp": "&", "lt": "<", "gt": ">", "quot": "\"", "apos": "'",
        "nbsp": "\u{00A0}", "hellip": "…", "mdash": "—", "ndash": "–",
        "rsquo": "’", "lsquo": "‘", "rdquo": "”", "ldquo": "“", "times": "×",
    ]
}

File: android/core/src/main/kotlin/com/automattic/wordpresstv/core/data/Html.kt
package com.automattic.wordpresstv.core.data

/**
 * Minimal HTML-to-plain-text helper. Strips tags and decodes the handful of
 * entities WP.com titles and excerpts actually contain (named + numeric, decimal
 * + hex). Mirrors the Apple `HTML` helper.
 */
internal object Html {
    private val tagRegex = Regex("<[^>]+>")

    fun plainText(raw: String): String = decodeEntities(stripTags(raw)).trim()

    fun stripTags(s: String): String = s.replace(tagRegex, "")

    fun decodeEntities(s: String): String {
        if (!s.contains('&')) return s
        val result = StringBuilder()
        var i = 0
        while (i < s.length) {
            val c = s[i]
            if (c == '&') {
                val semicolon = s.indexOf(';', startIndex = i + 1)
                if (semicolon != -1 && semicolon - i <= 12) {
                    val body = s.substring(i + 1, semicolon)
                    val decoded = decode(body)
                    if (decoded != null) {
                        result.append(decoded)
                        i = semicolon + 1
                        continue
                    }
                }
            }
            result.append(c)
            i++
        }
        return result.toString()
    }

    /** [body] is the entity without the leading `&` or trailing `;`. */
    private fun decode(body: String): String? {
        if (body.startsWith("#x") || body.startsWith("#X")) {
            val code = body.substring(2).toIntOrNull(16) ?: return null
            return codePoint(code)
        }
        if (body.startsWith("#")) {
            val code = body.substring(1).toIntOrNull() ?: return null
            return codePoint(code)
        }
        return named[body]
    }

    private fun codePoint(code: Int): String? =
        if (code in 0..0x10FFFF) String(Character.toChars(code)) else null

    private val named: Map<String, String> = mapOf(
        "amp" to "&", "lt" to "<", "gt" to ">", "quot" to "\"", "apos" to "'",
        "nbsp" to " ", "hellip" to "…", "mdash" to "—", "ndash" to "–",
        "rsquo" to "’", "lsquo" to "‘", "rdquo" to "”", "ldquo" to "“", "times" to "×",
    )
}
